lib/AST/ASTPrinter.cpp

```diff
@@ -3586,19 +3590,99 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
     return T->hasSimpleTypeRepr();
   }
 
+  /// Computes the map that is cached by `getVisibleClangModules()`.
+  /// Do not call directly.
+  llvm::DenseMap<const clang::Module *, ModuleDecl *>
+  computeVisibleClangModules() {
+    assert(Options.CurrentModule &&
+           "CurrentModule needs to be set to determine imported Clang modules");
+
+    llvm::DenseMap<const clang::Module *, ModuleDecl *> Result;
+
+    // For the current module, consider both private and public imports.
+    ModuleDecl::ImportFilter Filter = ModuleDecl::ImportFilterKind::Public;
+    Filter |= ModuleDecl::ImportFilterKind::Private;
+    SmallVector<ModuleDecl::ImportedModule, 4> Imports;
+    Options.CurrentModule->getImportedModules(Imports, Filter);
```

Comment:
I think the `SPIAccessControl` and `ShadowedBySeparateOverlay` import kinds also need to be included here, as those Clang modules will be visible. cc @xymus @brentdax

Reply:
`ShadowedBySeparateOverlay`, at least, should *not* be necessary; those modules will be visible transitively through the overlay's `@_exported import`.

Reply:
Right, this may change soon but for now `SPIAccessControl` imports should likely be included for the search from the current module.

Reply:
We've added a test for SPIAccessControl, and we fixed ASTPrinter to visit @_spi modules. Thanks for the suggestion!